### PR TITLE
[DOC, MRG] Fix CT alignment instructions

### PR DESCRIPTION
**Pull request by alexrockhill**
So in using the `Dipy` alignment code, I had been running into issues on some subjects as had others as mentioned on the discussion board. Per the issue in `Dipy` https://github.com/dipy/dipy/issues/2490 there is really no good solution for this despite several attempts. The issue with manual alignment is that it isn't that exact (in ``freeview`` you can only move in 0.5 mm increments with the sliders but you can type to any accuracy). So, I think, since this is really a situation where you want to be as exact as possible, the algorithm needs to be rerun with the manual starting point so that you don't get stuck in the local minimum.

I tried just using the aligned CT as a starting point, but, interestingly, that gives poor alignment results. I'm not quite sure why, but I'd really rather not try and diagnose the `Dipy` stuff as the Parzen histograms are cythonized and quite complex.

The solution I've found, and I think is better because it doesn't have the extra, intermediate cubic interpolation from the first alignment, is to use the `starting_affine` from the manual alignment. Unfortunately, you can't just get these from the freesurfer lta files as far as I know. I tried pulling out the array with

```
with open('path/to/ct.mgz.lta', 'r') as fid:
     reg_affine_fix = np.loadtxt(fid.readlines()[5:9])
```

But that's not the right affine and I'm not sure why. If anyone knows what the fix is for this, that would save the unideal `CT_orig` to `CT_aligned` step, which is a bit of a waste of computation.

EDIT: I've also used `zooms=dict(translation=5)` which speeds the above step up as well as the CT-T1 alignment up so that it takes half as long with the same results per my experience.

Anyway, by finding the `reg_affine` that `Dipy` uses to get from the original to the manually aligned CT, you can then use that as the starting alignment and solve to good precision without issue.

The main point is that, this algorithmically-aligned CT is a better approach in my opinion at a step that is very important.

### CODE REVIEW

**Comment by larsoner** [comment]
> But that's not the right affine and I'm not sure why

Have you tried posting a minimal example (original file, what you did, .lta file, then Python code that fails to resample properly) to the FS listserv to see if anyone can help?

My guess is that it's some problem with coordinate systems, like FS surface RAS vs scanner RAS, and/or an inversion needed somewhere...

**Comment by alexrockhill** [comment]
> > But that's not the right affine and I'm not sure why
> 
> Have you tried posting a minimal example (original file, what you did, .lta file, then Python code that fails to resample properly) to the FS listserv to see if anyone can help?
> 
> My guess is that it's some problem with coordinate systems, like FS surface RAS vs scanner RAS, and/or an inversion needed somewhere...

I haven't yet, I just today found something that worked finally. The rotations were different too, so I don't think it's that particular coordinate transform.

This is what was in the lta file

```
array([[ 0.99221027, -0.05494503,  0.11180324, -3.84350586],
       [ 0.05233596,  0.99828744,  0.02614108, -9.77523804],
       [-0.11304809, -0.02008611,  0.99338663, 15.25457001],
       [ 0.        ,  0.        ,  0.        ,  1.        ]])
```

For this actual manual alignment that worked for `Dipy`

```
array([[ 0.9922103 ,  0.11304873,  0.05233261,  2.91353137],
       [-0.11180388,  0.99338636, -0.02614255,  2.27159046],
       [-0.05494188,  0.02008791,  0.99828747, -0.02283438],
       [ 0.        ,  0.        ,  0.        ,  1.        ]])
```

If you can figure out how those two affines are related, you would once again out-puzzle me @larsoner :)

**Comment by alexrockhill** [comment]
You could also use the example data from the `Dipy` issue to figure it out as well if you have time.

I really just don't know and I'd like to get help from the Freesurfer folks but this only takes an extra ~5 minutes of computation (no executed in the tutorial) so I don't think it's critical right now.

**Comment by alexrockhill** [comment]
https://41345-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/clinical/10_ieeg_localize.html

I accidentally had the list in the code block though, should be fixed now.

**Comment by alexrockhill** [comment]
Ok, looks good https://41347-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/clinical/10_ieeg_localize.html

**Comment by larsoner** [comment]
From https://surfer.nmr.mgh.harvard.edu/fswiki/FsTutorial/LtaFormat :

> The .lta format is from subject ras to talairach ras, and stores affine.

Does that help? I'm still not totally clear on what you did with FreeSurfer to generate your LTA file. Was it part of recon-all, did you use freeview and export it, or...?

**Comment by larsoner** [comment]
Oh I see from reading the PR I think... I would expect this LTA to probably need to be multiplied by the inverse of the `subjects/sample/mri/transforms/talairach.lta` file. Empirically it looks like hopefully for that subject the values are similar to this (or its inverse or transpose or something):
```
import numpy as np
lta = np.array([[ 0.99221027, -0.05494503,  0.11180324, -3.84350586],
       [ 0.05233596,  0.99828744,  0.02614108, -9.77523804],
       [-0.11304809, -0.02008611,  0.99338663, 15.25457001],
       [ 0.        ,  0.        ,  0.        ,  1.        ]])
dipy = np.array([[ 0.9922103 ,  0.11304873,  0.05233261,  2.91353137],
       [-0.11180388,  0.99338636, -0.02614255,  2.27159046],
       [-0.05494188,  0.02008791,  0.99828747, -0.02283438],
       [ 0.        ,  0.        ,  0.        ,  1.        ]])
xform = np.linalg.solve(lta, dipy)
np.testing.assert_allclose(lta @ xform, dipy)
print(xform)
```
gives
```
[[  0.98484082   0.16188701  -0.06229771   9.06196382]
 [ -0.16502583   0.98506995  -0.0490249   11.961793  ]
 [  0.05343113   0.05856245   0.99685267 -14.10598892]
 [  0.           0.           0.           1.        ]]
```
Do you want to look further?

**Comment by alexrockhill** [comment]
Wait what? You just used the transform that you get from Freesurfer and combined it with the correct affine from the dipy alignment of the two CTs? I don't follow...

Regardless, I think the narrative as is in this PR makes sense and even if it saves 5 minutes of computation, inverting Talairach transforms and applying transforms to the Freesurfer affine isn't worth having to explain what's going on in the tutorial, especially since it's not actually executed. My vote is to stick with the simpler narrative that is computationally inefficient but easy to explain.

**Comment by larsoner** [comment]
Okay I did a (not very good) manual alignment, but instead of saving the modified volume, I did `Save reg...` to save out `reg.lta`, then clicked `Restore original`, then exited the transform dialog. I can then right-click the `sample_seeg_CT` and do "Apply transformation..." and it correctly applies it to get back to my manual alignment. So the information exists to directly use the manual alignment! To me if you do a manual alignment, and the information exists to use it, then you should use it directly rather than do any computation like currently done to re-figure-out the affine etc.

The following code seems to work, in the sense that `mri_vol2vol` and `FreeView` + `Apply transformation...` appears to give the same result:
```
# replicate behavior of
# $ mri_vol2vol --targ sample_seeg/mri/T1.mgz --mov sample_seeg_CT.mgz --lta reg.lta --o sample_seeg_CT_aligned.mgz
# which replicates the behavior of Freeview's Load Volume followed by Apply Transform

import numpy as np
import nibabel as nib
from dipy.align import resample

ctvox_t1vox_t = list()
with open('reg.lta', 'r') as fid:
    for li, line in enumerate(fid):
        if li in (5, 6, 7, 8):
            ctvox_t1vox_t.append(line.strip().split())
ctvox_t1vox_t = np.array(ctvox_t1vox_t, float)
ct_orig = nib.load('sample_seeg_CT.mgz')
T1 = nib.load('sample_seeg/mri/T1.mgz')
ct_aligned = resample(
    moving=np.array(ct_orig.dataobj),
    moving_affine=T1.affine @ ctvox_t1vox_t,
    static=np.array(T1.dataobj),
    static_affine=T1.affine)
nib.save(ct_aligned, 'sample_seeg_CT_aligned.mgz')
```
So I think this gets us most of the way there. @alexrockhill want to see if you can replicate this? And if it works, see if we can more directly utilize the LTA file rather than resample-then-computationally-figure-out-what-the-manual-transform-was?

**Comment by larsoner** [comment]
... or, to put it in `between_affine` form, which is cleaner:

<details>

```
import numpy as np
import nibabel as nib
from dipy.align import resample
np.set_printoptions(precision=4, suppress=True)

# replicate
# $ mri_vol2vol --targ sample_seeg/mri/T1.mgz --mov sample_seeg_CT.mgz --lta reg.lta --o sample_seeg_CT_aligned.mgz
# ...
# Vox2Vox Matrix is:
#  2.26969   0.78450   0.19857  -106.88104;
# -0.30369   0.27766   2.37424  -58.66954;
#  0.37370  -1.12662   0.17955   181.71855;
#  0.00000   0.00000   0.00000   1.00000;

ctvox_t1vox_t = list()
with open('reg.lta', 'r') as fid:
    for li, line in enumerate(fid):
        if li in (5, 6, 7, 8):
            ctvox_t1vox_t.append(line.strip().split())
ctvox_t1vox_t = np.array(ctvox_t1vox_t, float)
ct_orig = nib.load('sample_seeg_CT.mgz')
T1 = nib.load('sample_seeg/mri/T1.mgz')
# For dipy we need a mapping from T1 ras to CT ras, so we get there by going
# (right to left in these dot products):
# T1 ras -> T1 vox -> CT vox -> CT ras
t1ras_ctras_t = ct_orig.affine @ np.linalg.inv(ctvox_t1vox_t) @ np.linalg.inv(T1.affine)
print('Final Vox2Vox (T1 to CT) matrix is:')
print(np.linalg.inv(ctvox_t1vox_t))
# [[   2.2697    0.7845    0.1986 -106.881 ]
# [  -0.3037    0.2777    2.3742  -58.6695]
# [   0.3737   -1.1266    0.1796  181.7185]
# [   0.        0.        0.        1.    ]]
ct_aligned = resample(
    moving=np.array(ct_orig.dataobj),
    moving_affine=ct_orig.affine,
    static=np.array(T1.dataobj),
    static_affine=T1.affine,
    between_affine=t1ras_ctras_t)
print(np.array(ct_aligned.dataobj).max())
nib.save(ct_aligned, 'sample_seeg_CT_aligned_py.mgz')
```

</details>

Not 100% sure I have the names here all correct, but at least overlaying the results in FreeView I get the same numerical values (up to precision) for the `_py.mgz` and for the one using `mri_vol2vol`.

**Comment by alexrockhill** [comment]
Okay, I follow you now and that seems reasonable that it's some matrix multiplication but your code gives the aligned CT and what we need is the `reg_affine` or `between_affine` (in `resample`) because it's used as the `starting_affine` in the registration. I'm not exactly sure how to get that from the `moving_affine` + `static_affine` combo that you have above...

File: tutorials/clinical/10_ieeg_localize.py [hunk 1]
```diff
@@ -202,26 +202,43 @@ def plot_overlay(image, compare, title, thresh=None):
...
+#     .. code-block:: bash
+#
+#         $ freeview $MISC_PATH/seeg/sample_seeg/mri/T1.mgz \
+#            $MISC_PATH/seeg/sample_seeg_CT.mgz:colormap=heat:opacity=0.6
+#
+#     - Navigate to the upper toolbar, go to
+#       :menuselection:`Tools --> Transform Volume`
+#     - Use the rotation and translation slide bars to align the CT
+#       to the MR (be sure to have the CT selected in the upper left menu)
+#     - Save the modified volume using the ``Save Volume As...`` button
```
[the remark this answers is not in the document]
Comment (alexrockhill) [reply tutorials/clinical/10_ieeg_localize.py:217]: Yeah, my mistake, you needed to load it in and then align the original CT to the aligned version in the computationally inefficient version

**Comment by larsoner** [comment]
> I'm not exactly sure how to get that from the moving_affine + static_affine combo that you have above...

Did you see the code [up here in details](https://github.com/mne-tools/mne-python/pull/10298#issuecomment-1031798594)? It uses `between_affine`

**Comment by alexrockhill** [comment]
Ah sorry, no I didn't see that, but also I just checked and it gives the wrong answer...

```
CT_orig.affine @ np.linalg.inv(manual_reg_affine) @ np.linalg.inv(T1.affine)
array([[   0.41176715,    0.04691494,    0.02171942,  -51.41539919],
       [   0.02280218,   -0.00833573,   -0.41428914, -157.58287815],
       [  -0.09313208,    0.82749071,   -0.02177551,  -47.82375212],
       [   0.        ,    0.        ,    0.        ,    1.        ]])
```

where `manual_reg_affine` is

```
array([[ 0.99221027, -0.05494503,  0.11180324, -3.84350586],
       [ 0.05233596,  0.99828744,  0.02614108, -9.77523804],
       [-0.11304809, -0.02008611,  0.99338663, 15.25457001],
       [ 0.        ,  0.        ,  0.        ,  1.        ]])
```

And the correct output is

```
array([[ 0.9922103 ,  0.11304873,  0.05233261,  2.91353137],
       [-0.11180388,  0.99338636, -0.02614255,  2.27159046],
       [-0.05494188,  0.02008791,  0.99828747, -0.02283438],
       [ 0.        ,  0.        ,  0.        ,  1.        ]])
```

**Comment by alexrockhill** [comment]
Wait sorry, in my example, I used the alignment from the Dipy algorithm as a starting point, you figured it out, my bad

**Comment by alexrockhill** [comment]
Ok, sorry I was being dramatic, that wasn't as complicated as I thought it would be. I ran through the tutorial and it worked just fine, it does take ~5 minutes of computation on my Mac laptop so I think better to leave the execution out of the tutorial.

**Comment by alexrockhill** [comment]
Hmm, I'm not sure what's going on with the failures, doesn't seem related. Otherwise, I think ready to go.

File: tutorials/clinical/10_ieeg_localize.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (larsoner) [reply tutorials/clinical/10_ieeg_localize.py:217]: This needs to updated, it says to `Save Volume As...` but in that alignment dialog it's "Save registration" or something similar

[the remark this answers is not in the document]
Comment (alexrockhill) [reply tutorials/clinical/10_ieeg_localize.py:217]: If you save the volume it automatically saves the lta. I explained it clearly now

[the remark this answers is not in the document]
Comment (larsoner) [reply tutorials/clinical/10_ieeg_localize.py:217]: Why bother saving the volume? It's not used. I would just save the .lta then use it

**Comment by alexrockhill** [comment]
Just out of curiosity, @larsoner, why is it matrix multiplication in this case and not a dot product?

File: tutorials/clinical/10_ieeg_localize.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexrockhill) [reply tutorials/clinical/10_ieeg_localize.py:217]: Point taken

**Comment by larsoner** [comment]
> why is it matrix multiplication in this case and not a dot product?

I'm not sure what you mean... you mean why `a @ b` instead of `np.dot(a, b)`? These do the same thing for 2D arrays

**Comment by alexrockhill** [comment]
Oh gotcha, not sure why I was thinking `@` was `np.matmul` for a minute, that doesn't make sense as an output of the same shape... thanks for explaining

**Comment by alexrockhill** [comment]
Whew those deprecation warnings took a couple tries to filter, looks good now though https://41484-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/clinical/10_ieeg_localize.html

**Comment by larsoner** [comment]
Thanks @alexrockhill !

**Pull request merged**